File: engine/engine.js
class Engine {

  constructor(entities, config = {}) {
    this.entities = entities;
  }

  init() {    
    console.log('Initializing Engine');
  }

  start() {
    if ( this.running ) return;
    console.log('Starting Engine');
  }

  update(ctx) {
    // console.log('Update');
  }

  fixedUpdate(ctx) {
    // console.log('Fixed Update');
    this.entities.forEach(entity => {
      if (entity.hasOwnProperty('update')) {
        entity.update.call(entity, ctx);
      }
    });
  }

  stop() {
    console.log('Stopping Engine');
  }

  deinit() {
    console.log('Deinitializing Engine');
  }

}

export default Engine;

File: engine/engine.cpp
#include "engine.h"

#include "core/system/context.h"
#include "entity/context.h"
#include "entity/manager.h"
#include "entity/component.h"

#define LIFECYCLE_FUNC_IMPL(Name) \
void Engine::Name(SystemContext* context) { \
  EntityContext entityContext(context->time(), context->deltaTime()); \
  EntityManager* manager = context->get<EntityManager>(); \
  for (auto& entity : *manager) { \
    entity->Name(&entityContext); \
  } \
} \

namespace wage {

  // LIFECYCLE_FUNC_IMPL(start)
  // LIFECYCLE_FUNC_IMPL(update)
  // LIFECYCLE_FUNC_IMPL(fixedUpdate)
  // LIFECYCLE_FUNC_IMPL(stop)

}
